### ISSUE TITLE
Model weights are not fully set

### ISSUE THREAD

**Issue by Corendos** [facebookresearch/deepcluster#41]
Hi,

I've been wanting to use the trained models for a clustering task but I've come across a strange behavior.

The downloaded weights do not fully set the model weights. Between two successive runs, the output feature vector are different for the same input image.

Did I miss something ?

Thank you for your time

**Comment by mathildecaron31** [comment]
Hi,

Can you provide a minimal code to reproduce this ?

If the model is in [train mode](https://pytorch.org/docs/stable/nn.html#torch.nn.Module.train) it is expected that the feature vector differs for two successive runs because of Dropout layers. You need to do `model.eval()` to go in eval mode.

**Comment by Corendos** [comment]
Yeah, sure

Here is the code I used :
```
import argparse
import time
import os
import numpy as np
import torch
import torch.nn as nn
import torch.utils.data
import torchvision.datasets as datasets
import torchvision.transforms as transforms
from util import    AverageMeter
from models import alexnet, vgg16

class DeepClusterizer:
    def __init__(self):
        self.model = vgg16(sobel=True)
        self.model.top_layer = None
        self.model.classifier = nn.Sequential(*list(self.model.classifier.children())[:-1])
        self.model.cuda()
        self.features = nn.DataParallel(self.model.features)
    
    def compute_features(self, dataloader, N, batch_size):
        batch_time = AverageMeter()
        end = time.time()
        self.model.eval()
        # discard the label information in the dataloader
        for i, (input_tensor, _) in enumerate(dataloader):
            with torch.no_grad():
                input_var = torch.autograd.Variable(input_tensor.cuda())
            aux = self.model(input_var).data.cpu().numpy()

            if i == 0:
                features = np.zeros((N, aux.shape[1])).astype('float32')

            if i < len(dataloader) - 1:
                features[i * batch_size: (i + 1) * batch_size] = aux.astype('float32')
            else:
                # special treatment for final batch
                features[i * batch_size:] = aux.astype('float32')

            # measure elapsed time
            batch_time.update(time.time() - end)
            end = time.time()

            if i % 20 == 0:
                print('{0} / {1}\t'
                    'Time: {batch_time.val:.3f} ({batch_time.avg:.3f})'
                    .format(i, len(dataloader), batch_time=batch_time))
        return features
    
    def load_model(self, weights_filename):
        if os.path.isfile(weights_filename):
            print("=> loading checkpoint '{}'".format(weights_filename))
            checkpoint = torch.load(weights_filename)

            self.model.load_state_dict(checkpoint['state_dict'], False)
            print("=> loaded checkpoint '{}' (epoch {})"
                .format(weights_filename, checkpoint['epoch']))
        else:
            print("=> no checkpoint found at '{}'".format(weights_filename))
            raise RuntimeError()


def main():
    parser = argparse.ArgumentParser(description='DeepCluster test')

    parser.add_argument('data', metavar='DATASET_PATH', help='path to dataset')
    parser.add_argument('weights', metavar='WEIGHTS', help='path to model weights')
    parser.add_argument('--batch', default=256, type=int,
                        help='mini-batch size (default: 256)')
    args = parser.parse_args()
    
    normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406],
                                     std=[0.229, 0.224, 0.225])
    transform = [transforms.Resize(256),
                 transforms.CenterCrop(224),
                 transforms.ToTensor(),
                 normalize]

    dataset = datasets.ImageFolder(args.data, transform=transforms.Compose(transform))

    dataloader = torch.utils.data.DataLoader(dataset,
                                             batch_size=args.batch,
                                             num_workers=4,
                                             pin_memory=True)
        

    deepclusterizer = DeepClusterizer()
    deepclusterizer.load_model(args.weights)
    features = deepclusterizer.compute_features(dataloader, len(dataset), args.batch)

    print(features)


if __name__ == "__main__":
    main()
```

It is a simplified version of what I am really using but that should produce different feature vectors for the same input dataset.

Thank you for your reply !

**Comment by mathildecaron31** [comment]
I think you are not loading the model properly and you don't realize it because flag `strict` in `load_state_dict` is set to `False`.
Thus you end up with a randomly initialized model (which differ at each run if you don't set the random seed) instead of the deepcluster model.

Here is an example to load the model:
`
from util import load_model
self.model = load_model(weights_filename)

self.model.top_layer = None                                                  
self.model.classifier = nn.Sequential(*list(self.model.classifier.children())[:-1])
self.model.cuda()                                                           
self.features = nn.DataParallel(self.model.features)                    
`

**Comment by Corendos** [comment]
Hi,

I used the portion of code found here [main.py#L96](https://github.com/facebookresearch/deepcluster/blob/d85bd9d678c2876d19335bfaa834ef27bcba6d9b/main.py#L96) to load the model and had to do some modification, because it were not working properly.

Now it works, thanks for your help !
